src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -138,6 +141,27 @@ pub fn codegen_intrinsic_call<'a, 'tcx>(bx: &Builder<'a, 'tcx>,
             let llfn = cx.get_intrinsic(&("llvm.debugtrap"));
             bx.call(llfn, &[], None)
         }
+        "va_arg" => {
+            match fn_ty.ret.layout.abi {
+                layout::Abi::Scalar(ref scalar) => {
+                    match scalar.value {
+                        layout::Int(..) | layout::F64 | layout::Pointer => {
+                            bx.va_arg(args[0].immediate(), llret_ty)
+                        }
+                        // `va_arg` should never be used with the return type f32. If the
+                        // return type is f32, use `va_arg` with the return type of f64
+                        // followed by fptrunc to truncate the result to an f32.
+                        layout::F32 => {
+                            let ext_ret_ty = bx.va_arg(args[0].immediate(), Type::f64(cx));
+                            bx.fptrunc(ext_ret_ty, Type::f32(cx))
+                        }
```

Comment:
You can remove this special-casing here now.

Reply:
Removed. I was going to leave it for when arbitrary types are supported, but I think a `bug` here is probably better.